controllers/config/defaults.go

```diff
@@ -178,6 +178,8 @@ const DefaultMaxConcurrentReconciles = 10
 
 const DefaultRequeueTime = time.Second * 20
 
+var DSPAdeleted = make(map[string]bool)
```

Comment:
I have a few question:
* Are we going to store all DSPA forever? 
* How about concurrent deletion and creation using the same name? How the code is handling that?

Reply:
>Are we going to store all DSPA forever?

We are storing just to compare in the future if the same DSPA name gets created which basically tackles your second question.

>How about concurrent deletion and creation using the same name? How the code is handling that?

Line 185 and 186 in `dspipeline_controller.go` is specifically created to tackle this condition.

Reply:
May we have an unit test for this change?

Reply:
I don't think this is the right approach for this. Storing this sort of information memory is not scalable, and even though we can store a lot of dspa names, we would just lose that information upon a pod restart. Concurrency can be another issue as diego pointed out in the future if we scaled the operator. 


the offending piece of code _should probably be_  here: 

https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/controllers/dspipeline_controller.go#L531-L537

Essentially we requeue after the cr has been deleted because of pod changes. If we can find a way to not queue after a cr is deleted that would be great, this would be the optimal approach. Have you considered this? if so, I'd be curious to know why we couldn't go this route.

Reply:
@HumairAK Unfortunately, reading through there is no such way as the `SetupManager` still looks for the said resources as it belongs to the CR after deletion and is being watched by the Controller as specified (https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/controllers/dspipeline_controller.go#L507)

I tried using the predicates mentioned here (https://github.com/kubernetes-sigs/kubebuilder/issues/1700#issuecomment-707063707) to see if that helped filter out the Resources which are already deleted like the code snippet mentioned below but that doesn't help as mentioned in the Operator-SDK doc below.

>This is an implementation of the controller that simply filters Delete events on Pods that have been confirmed deleted; the controller receives all Delete events that occur, and we may only care about resources that have not been completely deleted

```
func ignoreDeletionPredicate() predicate.Predicate {
	return predicate.Funcs{
		UpdateFunc: func(e event.UpdateEvent) bool {
			// Ignore updates to CR status in which case metadata.Generation does not change
			return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
		},
		DeleteFunc: func(e event.DeleteEvent) bool {
			// Evaluates to false if the object has been confirmed deleted.
			return !e.DeleteStateUnknown
		},
	}
}

func (r *DSPAReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&dspav1alpha1.DataSciencePipelinesApplication{}).
		Owns(&appsv1.Deployment{}).
		....
		....
		....
		WithEventFilter(ignoreDeletionPredicate()).
		Complete(r)
}

```

But reading through your comment, instead of creating a separate variable to store in memory with the flag, is there any harm in storing a bool flag in the DataSciencePipelinesApplication struct itself here (https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/api/v1alpha1/dspipeline_types.go#L404) and leverage the same to stop reconcilation in `dspipeline_controller.go`

Reply:
Well this happens when dspa is deleted, so adding a bool to dspa is not going to help I think. 

Another thought I had was utilizing finalizers for this. Having DSPA only remove the finalizer when all pods are cleaned up, but I don't think it makes sense to block DSPA deletion on that, it would have an unwanted impact to deletion times in the UI. 

Since we are already exiting reconcile early if we detect DSPA is not found and not reporting an error. I'm thinking we just convert the "DSPA not found" log to a debug, and just disable debug logs by default (we should be doing this anyway). 

I would like to hear others' thoughts on this approach though. @gregsheremeta  or @gmfrasca any thoughts?

Reply:
Finalizers is the correct approach.

1 kube delete is called on DSPA. deletionTimestamp is set
2 UI treats any DSPA with a deletionTimestamp as deleted. "Pipeline Server" is no longer rendered almost immediately.
3 Resources are cleaned up.
4 resources in finalizer are gone -> finalizer is removed -> DSPA is actually deleted from kube apiserver

I've never coded this myself, so I might have this a litte imperfect, but this is  roughly the pattern I've seen in production kube systems I've worked on.

Reply:
> this is roughly the pattern I've seen in production kube systems I've worked on

I do concede, though, that this was quite the nightmare, because things were constantly getting stuck

Reply:
> Since we are already exiting reconcile early if we detect DSPA is not found and not reporting an error. I'm thinking we just convert the "DSPA not found" log to a debug, and just disable debug logs by default (we should be doing this anyway).

Ok, yeah, I'm sold. I guess this drastically changes this PR -- sorry Achyut :)

so l ine 180 below, change to roughly

```
log.Debug("DSPA resource was not found, assuming it was recently deleted, nothing to do here")
```

and then set DSPO default to debug logging (probably in a different PR :smile:)